### PR TITLE
GH-50515: [C++][Compute] Respect parent validity bitmap when casting nested structs with non-nullable fields

### PR DESCRIPTION
**Pull request by aaron-seq**
### Rationale for this change

When casting a nested struct whose inner field changes from nullable to
non-nullable, `CastStruct::Exec` currently calls
`in_values->GetNullCount() > 0` without considering the parent struct's
validity bitmap. This causes a spurious `ArrowInvalid` error when the
child's physical nulls are fully masked by null parent entries.

Reported in #50515.

### What changes are included in this PR?

In `scalar_cast_nested.cc`, replace the unconditional rejection with a
three-way check:

1. **Parent has no bitmap** → child nulls are unmasked; reject.
2. **Child has no bitmap** → element-wise check against parent validity.
3. **Both bitmaps present** → use `BinaryBitBlockCounter::NextAndNotWord()`
   to detect parent-valid-AND-child-null positions without heap allocation.

### Are these changes tested?

Five new C++ test cases in `scalar_cast_test.cc`:
- `StructNestedNullabilityAbsentParent` — true violation, must reject
- `StructNestedNullabilityMasked` — masked null, must succeed
- `StructNestedNullabilitySliced` — offset correctness for both outcomes
- `StructNestedNullabilityAbsentChild` — no nulls, must succeed
- `StructNestedNullabilityDeep` — 3-level nesting with masked null

One new Python test in `test_compute.py`:
- `test_cast_struct_nested_nullability` — end-to-end PyArrow validation
  covering success, rejection, and sliced array cases

### Are there any user-facing changes?

Struct casts that were previously rejected with
`ArrowInvalid: field '...' has nulls` will now succeed when the nulls
are masked by parent-level nulls. This is a correctness fix, not a
behavior change — the previous behavior was incorrect per the Arrow
columnar format specification.

* GitHub Issue: #50515

### CODE REVIEW

**Comment by pitrou** [comment]
>     3. **Both bitmaps present** → use `BinaryBitBlockCounter::NextAndNotWord()`
>        to detect parent-valid-AND-child-null positions without heap allocation.

You can use `CountAndSetBits` which does exactly that :)

File: cpp/src/arrow/compute/kernels/scalar_cast_nested.cc [hunk 1]
```diff
@@ -384,12 +385,48 @@ struct CastStruct {
...
+            int64_t position = 0;
+            while (position < in_array.length) {
+              arrow::internal::BitBlockCount block = bit_counter.NextAndNotWord();
+              if (block.popcount > 0) {
+                has_nulls = true;
+                break;
+              }
+              position += block.length;
+            }
+          }
```
Comment (pitrou) [inline cpp/src/arrow/compute/kernels/scalar_cast_nested.cc]: If we get here, I think we should make sure that the final casted child does not have a null bitmap, otherwise it might violate expectations for a nullable field (which are not well specified, unfortunately).

File: cpp/src/arrow/compute/kernels/scalar_cast_nested.cc [hunk 2]
```diff
@@ -384,12 +385,48 @@ struct CastStruct {
...
+          bool has_nulls = false;
+          const uint8_t* parent_bitmap = in_array.buffers[0].data;
+          const uint8_t* child_bitmap = in_values->buffers.empty() ? nullptr : (in_values->buffers[0] ? in_values->buffers[0]->data() : nullptr);
+
+          if (parent_bitmap == nullptr) {
+            // Parent has no nulls. Since child has nulls, they are unmasked.
+            has_nulls = true;
+          } else if (child_bitmap == nullptr) {
+            // Child has nulls but no bitmap (e.g. NullArray, RunEndEncoded, Union).
+            // We must semantically check if any valid parent element corresponds to a null child.
```
Comment (pitrou) [inline cpp/src/arrow/compute/kernels/scalar_cast_nested.cc]: The `BinaryBitBlockCounter` path below does not account for logical nulls, so I don't think this one should, either.

We can just hardcode an error for NullArray.

**Comment by pitrou** [comment]
@aaron-seq @davlee1972 thanks for the issue and PR.

This is a bit tricky as we have never formalized what the nullable flag means for non-trivial types, so I've started a discussion on the dev ML to make sure we don't go into the wrong direction here:
https://lists.apache.org/thread/osgpy5b12bhyb2vpxd8brtv9zzc4cq9s

**Comment by davlee1972** [comment]
I understand the complexity and for pure performance reasons I don't think you want to add Nulls in a non nullable array to begin with, but I'm forced to even out array lengths between parent records and child records.

I added a function called fill_nulls_with_dummy() as a workaround to populate any non nullable array with zeros, empty string, 1970 dates, 0 duration, etc..

Maybe what we really need is a pyarrow.notnull() type, an option in pyarrow.compute.fill_null() to use a dummy fill value based on the values type and a pyarrow compute non_null_scalar() function that takes a type and returns back a dummy scalar.

**Comment by davlee1972** [comment]
I actually like how assembling list arrays work with a pure non nullable array with an offsets array which has None for null positions without having to pollute the data array with None values.

It would be a major change to support an offsets array for struct arrays.. [0,1,2,None,None,3]. For a struct array with only four non nullable array values but contains two additional null entries.

**Comment by aaron-seq** [comment]
Thanks for the review! Pushed a follow-up commit addressing all three points:

CountAndSetBits: replaced the manual BinaryBitBlockCounter/NextAndNotWord loop with CountSetBits(parent) - CountAndSetBits(parent, child) for the case where both parent and child have validity bitmaps.
No-bitmap child (NullType/REE/Union): dropped the per-element logical-null check and now conservatively reject whenever the child reports any nulls and has no validity buffer of its own consistent with the bitmap-only semantics used in the main path, rather than mixing in logical-null reasoning.

Non-nullable field output: the cast result for a non-nullable field now has its validity buffer stripped before being attached to the parent. Previously the masked-but-physically-null bits from the source could survive into the output, which would surface as real nulls if that field were later extracted on its own (e.g. struct_field/flatten) thanks for catching that, it was the one actual correctness gap left.

Also added a regression test for the no-bitmap-child path (the existing "AbsentChild" test didn't actually exercise it, so I renamed it to reflect what it covers and added a new one), and fixed a couple of lint nits (an overlong line, a missing blank line in the Python test).

File: cpp/src/arrow/compute/kernels/scalar_cast_nested.cc [hunk 3]
```diff
@@ -384,17 +384,58 @@ struct CastStruct {
...
+            unmasked_null_count = in_values->GetNullCount();
+          } else {
+            // Both parent and child have bitmaps: an unmasked null is a position
+            // where the parent is valid but the child is null, i.e. parent bits
+            // set that aren't also set in the child.
+            unmasked_null_count = arrow::internal::CountSetBits(
+                                      parent_bitmap, in_array.offset, in_array.length) -
+                                  arrow::internal::CountAndSetBits(
+                                      parent_bitmap, in_array.offset, child_bitmap,
+                                      in_values->offset, in_array.length);
```
Comment (pitrou) [inline cpp/src/arrow/compute/kernels/scalar_cast_nested.cc]: Can we instead add a function `CountAndNotSetBits` in `arrow/util/bitmap_ops.h` and use it directly here?

File: cpp/src/arrow/compute/kernels/scalar_cast_nested.cc [hunk 4]
```diff
@@ -384,17 +384,58 @@ struct CastStruct {
...
+                  ? nullptr
+                  : in_values->buffers[0]->data();
+
+          int64_t unmasked_null_count;
+          if (parent_bitmap == nullptr) {
+            // Parent has no nulls, so any child null is unmasked.
+            unmasked_null_count = in_values->GetNullCount();
+          } else if (child_bitmap == nullptr) {
+            // Child reports nulls but has no validity buffer of its own (e.g.
+            // NullArray, RunEndEncoded, Union). There is no cheap, bitmap-only way
```
Comment (pitrou) [inline cpp/src/arrow/compute/kernels/scalar_cast_nested.cc]: This will only apply to NullArray. RunEndEncoded and Union should return 0 for `GetNullCount` (which should not be confused with `ComputeLogicalNullCount`).

File: cpp/src/arrow/compute/kernels/scalar_cast_test.cc [hunk 5]
```diff
@@ -4145,6 +4145,128 @@ TEST(Cast, StructToStructSubsetWithNulls) {
...
+TEST(Cast, StructNestedNullabilityMasked) {
+  auto inner_type_dest = struct_({field("a", int32(), /*nullable=*/false)});
+  auto outer_type_dest = struct_({field("inner", inner_type_dest)});
+  auto inner_type_src = struct_({field("a", int32())});
+  auto outer_type_src = struct_({field("inner", inner_type_src)});
+
+  auto src = ArrayFromJSON(outer_type_src, R"([
+    {"inner": {"a": 1}},
+    null
+  ])");
```
Comment (pitrou) [inline cpp/src/arrow/compute/kernels/scalar_cast_test.cc]: This does not guarantee that the inner field will have nulls at all, it's just an implementation detail of `ArrayFromJSON`. Can you instead use `StructArray::Make` to construct the outer array from its inner child array?

File: cpp/src/arrow/compute/kernels/scalar_cast_test.cc [hunk 6]
```diff
@@ -4145,6 +4145,128 @@ TEST(Cast, StructToStructSubsetWithNulls) {
...
+    {"inner": {"a": 1}},
+    {"inner": {"a": 2}}
+  ])");
+  CheckCast(src, expected);
+}
+
+TEST(Cast, StructNestedNullabilityNoChildBitmapConservativelyRejected) {
+  // NullType (and other child types without a validity buffer of their own,
+  // e.g. RunEndEncoded/Union) have no bitmap to distinguish masked from
+  // unmasked nulls, so any reported null is conservatively rejected -- even
```
Comment (pitrou) [inline cpp/src/arrow/compute/kernels/scalar_cast_test.cc]: I don't think REE or Union would be rejected, can you fix the comment?

File: cpp/src/arrow/compute/kernels/scalar_cast_test.cc [hunk 7]
```diff
@@ -4145,6 +4145,128 @@ TEST(Cast, StructToStructSubsetWithNulls) {
...
+  auto inner_type_src = struct_({field("a", int32())});
+  auto outer_type_src = struct_({field("inner", inner_type_src)});
+
+  auto src = ArrayFromJSON(outer_type_src, R"([
+    {"inner": {"a": 1}},
+    {"inner": {"a": null}}
+  ])");
+  EXPECT_RAISES_WITH_MESSAGE_THAT(
+      Invalid, ::testing::HasSubstr("has nulls. Can't cast to non-nullable field"),
+      Cast(src, CastOptions::Safe(outer_type_dest)));
```
Comment (pitrou) [inline cpp/src/arrow/compute/kernels/scalar_cast_test.cc]: Can you also check with a zero-length input?

File: cpp/src/arrow/compute/kernels/scalar_cast_test.cc [hunk 8]
```diff
@@ -4145,6 +4145,128 @@ TEST(Cast, StructToStructSubsetWithNulls) {
...
+
+  // Slicing to only the masked (fully-null outer) row is still rejected.
+  EXPECT_RAISES_WITH_MESSAGE_THAT(
+      Invalid, ::testing::HasSubstr("has nulls. Can't cast to non-nullable field"),
+      Cast(src->Slice(1, 1), CastOptions::Safe(outer_type_dest)));
+}
```
Comment (pitrou) [inline cpp/src/arrow/compute/kernels/scalar_cast_test.cc:4246]: Can you also check with a zero-length input? (not sure whether it should succeed or fail)

**Review by pitrou — COMMENTED** [review]
Thanks for the update @aaron-seq ! Here are a couple more suggestions.

**Comment by aaron-seq** [comment]
Thanks @pitrou, all five points addressed in da10b42.

**`CountAndNotSetBits`** — added to `arrow/util/bitmap_ops.h`/`.cc` (same `BinaryBitBlockCounter` shape as `CountAndSetBits`, using `NextAndNotWord()`), and the kernel now calls it directly instead of subtracting `CountAndSetBits` from `CountSetBits`. Added `BitUtilTests.TestCountAndNotSetBits` covering unaligned offsets on both sides, zero length, and the `(x & y) + (x & ~y) == x` identity.

**Bitmap-less children** — you're right, I had this backwards. `GetNullCount()` only counts physical nulls, so union and run-end encoded report 0 and never reach the check at all. NullType is the only type that gets there without a validity buffer, so that branch is now just part of the rejection condition, with the comment rewritten to say so.

**Test construction** — the nested-nullability inputs are now built with `StructArray::Make` from an explicit `"a"` child plus explicit validity bitmaps at each level, so where `"a"` physically holds nulls is pinned down by the test rather than being an artifact of `ArrayFromJSON`. This also caught that the sliced test wasn't exercising what its comment claimed: the `[3, 5)` slice had no null in `"a"` at all, so it was passing via the `GetNullCount() == 0` early exit rather than through the masked-null path. Fixed the fixture so it actually hits it.

**NullType test comment** — rewritten; it no longer claims REE/union are rejected, and notes they report no physical nulls so they're unaffected.

**Zero-length input** — added on both paths. It succeeds: an empty slice reports no nulls, so the check is skipped. Worth noting the NullType case is the interesting one — `src->Slice(1, 1)` is rejected but `src->Slice(1, 0)` succeeds, since `ArrayData::Slice` carries `null_count = len` for an all-null array.

Also stripped trailing whitespace in `test_compute.py`, which was the autopep8 lint failure. The two red R jobs are unrelated — both are `R CMD check` failing on a Bioconductor index fetch (`cannot open URL '.../PACKAGES'`), with `FAIL 0` on the actual test suites.

One thing I left alone pending the ML thread: the check runs per cast level, so if the outer element is null but the inner struct is *physically* valid there with a null leaf, the recursive inner cast still rejects it — it has no view of the outer bitmap. I didn't add a test asserting that either way, since it seems to be exactly what's under discussion.
